File: scripts/validate-config-naming.py
#!/usr/bin/env python3
"""Validate changed configuration identifiers against central naming rules."""

from __future__ import annotations

import re
import subprocess
import sys
from pathlib import Path

PLATFORM_PREFIXES = ("RUNNER_", "ACTIONS_")
PLATFORM_NAMES = {
    "GH_TOKEN",
    "NODE_OPTIONS",
    "GITHUB_ACTION",
    "GITHUB_ACTIONS",
    "GITHUB_ACTOR",
    "GITHUB_ACTOR_ID",
    "GITHUB_API_URL",
    "GITHUB_BASE_REF",
    "GITHUB_ENV",
    "GITHUB_EVENT_NAME",
    "GITHUB_EVENT_PATH",
    "GITHUB_GRAPHQL_URL",
    "GITHUB_HEAD_REF",
    "GITHUB_JOB",
    "GITHUB_OUTPUT",
    "GITHUB_PATH",
    "GITHUB_REF",
    "GITHUB_REF_NAME",
    "GITHUB_REF_PROTECTED",
    "GITHUB_REF_TYPE",
    "GITHUB_REPOSITORY",
    "GITHUB_REPOSITORY_ID",
    "GITHUB_REPOSITORY_OWNER",
    "GITHUB_REPOSITORY_OWNER_ID",
    "GITHUB_RETENTION_DAYS",
    "GITHUB_RUN_ATTEMPT",
    "GITHUB_RUN_ID",
    "GITHUB_RUN_NUMBER",
    "GITHUB_SERVER_URL",
    "GITHUB_SHA",
    "GITHUB_STEP_SUMMARY",
    "GITHUB_TRIGGERING_ACTOR",
    "GITHUB_WORKFLOW",
    "GITHUB_WORKFLOW_REF",
    "GITHUB_WORKFLOW_SHA",
    "GITHUB_WORKSPACE",
}
TOKEN_RE = re.compile(r"\b[A-Z][A-Z0-9]*(?:_[A-Z0-9]+)+\b")


def parts(name: str) -> list[str]:
    return [part for part in name.strip("_").split("_") if part]


def changed_paths(root: Path, base: str, head: str) -> list[Path]:
    proc = subprocess.run(
        ["git", "diff", "--name-only", "--diff-filter=ACMR", base, head],
        cwd=root,
        check=True,
        text=True,
        stdout=subprocess.PIPE,
    )
    return [Path(raw) for raw in proc.stdout.splitlines() if raw and (root / raw).is_file()]


def should_scan(path: Path) -> bool:
    if path.parts[:2] == (".github", "workflows") and path.suffix.lower() in {".yml", ".yaml"}:
        return True
    if path.name in {".env.variables", ".secrets.required", ".dev.vars.example"}:
        return True
    if path.name == "wrangler.jsonc":
        return True
    if path.parts and path.parts[0] == "config" and path.suffix.lower() in {".json", ".jsonc", ".toml", ".yml", ".yaml"}:
        return True
    return False


def is_platform_name(name: str) -> bool:
    return name in PLATFORM_NAMES or name.startswith(PLATFORM_PREFIXES)


def main() -> int:
    if len(sys.argv) != 3:
        print("usage: validate-config-naming.py <base-sha> <head-sha>", file=sys.stderr)
        return 64

    root = Path.cwd()
    errors: list[str] = []
    checked = 0

    for path in changed_paths(root, sys.argv[1], sys.argv[2]):
        if not should_scan(path):
            continue
        checked += 1
        text = (root / path).read_text(encoding="utf-8")
        for name in sorted(set(TOKEN_RE.findall(text))):
            if is_platform_name(name):
                continue
            if len(parts(name)) > 3:
                errors.append(f"{path.as_posix()}: configuration name '{name}' exceeds three segments")
            if name.startswith("GH_") and name != "GH_TOKEN":
                errors.append(f"{path.as_posix()}: custom GitHub configuration '{name}' must use GITHUB_")
            if name.startswith("CF_"):
                errors.append(f"{path.as_posix()}: custom Cloudflare configuration '{name}' must use CLOUDFLARE_")
            if name.endswith("_PAT") or "_PAT_" in name:
                errors.append(f"{path.as_posix()}: credential '{name}' must use TOKEN or KEY instead of PAT")

    if errors:
        print("configuration naming violations:", file=sys.stderr)
        for error in sorted(set(errors)):
            print(f"  {error}", file=sys.stderr)
        return 1

    print(f"configuration naming conventions passed ({checked} changed config files)")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/validate-config-naming.ts
import { readFile } from "node:fs/promises";
import { basename, extname } from "node:path";
import { CliError, handleError, isMain, runText } from "./runtime-command.ts";

const platformNames = new Set([
  "GH_TOKEN",
  "NODE_OPTIONS",
  "GITHUB_ACTION",
  "GITHUB_ACTIONS",
  "GITHUB_ACTOR",
  "GITHUB_ACTOR_ID",
  "GITHUB_API_URL",
  "GITHUB_BASE_REF",
  "GITHUB_ENV",
  "GITHUB_EVENT_NAME",
  "GITHUB_EVENT_PATH",
  "GITHUB_GRAPHQL_URL",
  "GITHUB_HEAD_REF",
  "GITHUB_JOB",
  "GITHUB_OUTPUT",
  "GITHUB_PATH",
  "GITHUB_REF",
  "GITHUB_REF_NAME",
  "GITHUB_REF_PROTECTED",
  "GITHUB_REF_TYPE",
  "GITHUB_REPOSITORY",
  "GITHUB_REPOSITORY_ID",
  "GITHUB_REPOSITORY_OWNER",
  "GITHUB_REPOSITORY_OWNER_ID",
  "GITHUB_RETENTION_DAYS",
  "GITHUB_RUN_ATTEMPT",
  "GITHUB_RUN_ID",
  "GITHUB_RUN_NUMBER",
  "GITHUB_SERVER_URL",
  "GITHUB_SHA",
  "GITHUB_STEP_SUMMARY",
  "GITHUB_TRIGGERING_ACTOR",
  "GITHUB_WORKFLOW",
  "GITHUB_WORKFLOW_REF",
  "GITHUB_WORKFLOW_SHA",
  "GITHUB_WORKSPACE",
]);

const tokenPattern = /\b[A-Z][A-Z0-9]*(?:_[A-Z0-9]+)+\b/g;
const workflowExternalPattern = /\$\{\{\s*(?:vars|secrets)\.([A-Z][A-Z0-9_]*)/g;
const selfIdentifyingPrefixes = [
  "AW_",
  "AIG_",
  "AI_GATEWAY_",
  "APP_SOURCE_",
  "DELTA_",
  "CLOUDFLARE_",
  "GITHUB_",
  "AWS_",
  "GCP_",
  "ALGOLIA_",
  "TAILSCALE_",
];

function isPlatformName(name: string): boolean {
  return platformNames.has(name) || name.startsWith("RUNNER_") || name.startsWith("ACTIONS_");
}

const ownerScopes = new Set(["projects", "apps", "services", "tools", "crates", "skills"]);

function scopedOwnerPrefix(path: string): string | null {
  const segments = path.split("/").filter(Boolean);
  for (let index = 0; index < segments.length - 1; index += 1) {
    const segment = segments[index] ?? "";
    if (!ownerScopes.has(segment)) continue;
    const owner = segments[index + 1] ?? "";
    const token = owner.replace(/[^A-Za-z0-9]+/g, "_").replace(/^_+|_+$/g, "").toUpperCase();
    return token ? `${token}_` : null;
  }
  return null;
}

function isSelfIdentifying(name: string, path: string): boolean {
  if (selfIdentifyingPrefixes.some((prefix) => name.startsWith(prefix))) {
    return true;
  }
  const ownerPrefix = scopedOwnerPrefix(path);
  return Boolean(ownerPrefix && name.startsWith(ownerPrefix));
}

function externalNames(path: string, text: string): string[] {
  if (path.startsWith(".github/workflows/") && [".yml", ".yaml"].includes(extname(path).toLowerCase())) {
    return [...new Set([...text.matchAll(workflowExternalPattern)].map((match) => match[1]!).filter(Boolean))].sort();
  }
  return [...new Set(text.match(tokenPattern) ?? [])].sort();
}

function shouldScan(path: string): boolean {
  if (path.startsWith(".github/workflows/") && [".yml", ".yaml"].includes(extname(path).toLowerCase())) {
    return true;
  }
  const name = basename(path);
  if ([".env", ".env.example", ".env.variables", ".secrets.required", ".dev.vars.example"].includes(name)) {
    return true;
  }
  if (name === "wrangler.jsonc") {
    return true;
  }
  return path.split("/").includes("config") && [".json", ".jsonc", ".toml", ".yml", ".yaml"].includes(extname(path).toLowerCase());
}

export function validateConfigText(path: string, text: string): string[] {
  const errors: string[] = [];
  const names = externalNames(path, text);

  for (const name of names) {
    if (isPlatformName(name)) {
      continue;
    }
    if (!isSelfIdentifying(name, path)) {
      errors.push(`${path}: external configuration '${name}' must identify its owning system without repository context`);
    }
    if (name.endsWith("_PAT") || name.includes("_PAT_")) {
      errors.push(`${path}: credential '${name}' must use TOKEN or KEY instead of PAT`);
    }
    const usesBooleanPrefix = /(?:^|_)(?:IS|HAS|CAN|SHOULD)_/.test(name);
    const isModeValue = name.endsWith("_MODE");
    const looksBoolean = !isModeValue && (
      /(?:^|_)(?:ALLOW|ENABLE|DISABLE|EXPOSE|INCLUDE)_/.test(name)
      || name.endsWith("_ENABLED")
    );
    if (looksBoolean && !usesBooleanPrefix) {
      errors.push(`${path}: Boolean configuration '${name}' must include an IS_, HAS_, CAN_, or SHOULD_ semantic segment`);
    }
  }
  return errors;
}

export async function validateConfigNames(base: string, head: string): Promise<number> {
  const changed = await runText("git", ["diff", "--name-only", "--diff-filter=ACMR", base, head]);
  const paths = changed ? changed.split(/\r?\n/).filter(Boolean).filter(shouldScan) : [];
  const errors: string[] = [];

  for (const path of paths) {
    try {
      const text = await readFile(path, "utf8");
      errors.push(...validateConfigText(path, text));
    } catch {
      // Deleted or unavailable files are excluded by --diff-filter and do not block validation.
    }
  }
  if (errors.length > 0) {
    for (const error of [...new Set(errors)].sort()) {
      console.error(`::error::${error}`);
    }
    return 1;
  }

  console.log(`Configuration naming conventions passed (${paths.length} changed config files).`);
  return 0;
}

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  if (args.length !== 2) {
    throw new CliError("Usage: validate-config-naming.ts <base-sha> <head-sha>", 64);
  }
  process.exitCode = await validateConfigNames(args[0] ?? "", args[1] ?? "");
}

if (isMain(import.meta.url)) {
  main().catch(handleError);
}
